crates/environ/src/module.rs

```diff
@@ -259,6 +259,100 @@ impl ModuleTranslation<'_> {
         }
         self.module.memory_initialization = MemoryInitialization::Paged { map };
     }
+
+    /// Attempts to convert the module's table initializers to
+    /// FuncTable form where possible. This enables lazy table
+    /// initialization later by providing a one-to-one map of initial
+    /// table values, without having to parse all segments.
+    pub fn try_func_table_init(&mut self) {
+        // This should be large enough to support very large Wasm
+        // modules with huge funcref tables, but small enough to avoid
+        // OOMs or DoS on truly sparse tables.
+        const MAX_FUNC_TABLE_SIZE: u32 = 1024 * 1024;
+
+        let num_imported_tables = self.module.num_imported_tables;
+        for (table_index, init) in &mut self.module.table_initialization {
+            // If this is an imported table, we can't provide a
+            // preconstructed table for it, because its values depend
+            // on the imported table overlaid with whatever segments
+            // we have.
+            if table_index.index() < num_imported_tables {
```

Comment:
I think most other places we check this by using `module.defined_table_index(index)`, so coul that be used here instead?

Reply:
Ah, it's a borrowing thing (we take a `&mut self.module.table_initialization` via the iterator); but I've moved this to a `.skip` on the iter, which is I think a pattern used elsewhere.

Reply:
Actually now with refactor this went away and we can use `defined_table_index` as you suggest.